### PR TITLE
feat(components): add `Image` component

### PR DESCRIPTION
**Pull request by EmilyZhang777**
This PR sets up the components/ export and adds an Image component.

J=SUMO-4687
TEST=manual

Ran against both dev and production builds on the starter repo, made sure I can import the Image component and all functionalities work as specified [here](https://www.notion.so/Images-c04403f8e8f14e7bb3869894d5a4860b).

### CODE REVIEW

File: packages/pages/src/components/image/types.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+  };
+};
+
+export type ImageProps = {
+  image: ImageType;
+  className?: string;
+  width?: number;
+  height?: number;
+  aspectRatio?: number;
+  layout?: "intrinsic" | "fixed" | "aspect" | "fill";
```
Comment (mancusi) [inline packages/pages/src/components/image/types.ts:24]: use a typescript enum for this

```ts
enum Layout = {
  INTRINSIC = "intrinsic"
  ...
}
```

File: packages/pages/src/components/image/types.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+export type ImageType = {
+  image: {
+    height: number;
+    width: number;
+    url: string;
+    thumbnails?: ThumbnailType[];
+  };
+};
+
+export type ImageProps = {
```
Comment (mancusi) [inline packages/pages/src/components/image/types.ts:18]: make sure to document all of these props

File: packages/pages/src/components/image/image.tsx [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+  const imgHeight: number = image.image.height;
+  const imgUUID: string = image.image.url.split("/")[4];
+
+  style.objectFit = style.objectFit || "cover";
+  style.objectPosition = style.objectPosition || "center";
+
+  let widths: number[] = [100, 320, 640, 960, 1280, 1920];
+  let src: string = getImageUrl(imgUUID, 500, 500);
+
+  switch (layout) {
```
Comment (mancusi) [inline packages/pages/src/components/image/image.tsx:41]: can you extract this to a separate function?

File: packages/pages/src/components/image/image.tsx [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+
+  if (layout != "fixed" && (width || height)) {
+    console.warn(
+      "Width or height is passed in but layout is not fixed. These will have no impact. If you want to have a fixed height or width then set layout to fixed."
+    );
+  }
+
+  const imgWidth: number = image.image.width;
+  const imgHeight: number = image.image.height;
+  const imgUUID: string = image.image.url.split("/")[4];
```
Comment (mancusi) [inline packages/pages/src/components/image/image.tsx:33]: can we validate that the image url is of the form we expect before doing this?

File: packages/pages/src/components/image/index.ts [hunk 5]
```diff
@@ -0,0 +1,2 @@
+export { Image } from "./image";
+export * from "./types";
```
[the remark this answers is not in the document]
Comment (EmilyZhang777) [reply packages/pages/src/components/image/index.ts:2]: I tested that this only exports the types that are denoted with "export". If we want to have an internal type, we can just not export it in types.ts

File: packages/pages/src/components/image/image.tsx [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+/**
+ * A component that renders an image. Here is an example of using the component with simple and
+ * complex image fields from Knowledge Graph:
+ * ```
+ * import { Image } from "@yext/pages/components";
+ *
+ * const template = ({ document }) => {
+ *   return (
+ *     <div>
+ *       <Image image={document.logo.image} />
```
Comment (Mxs2019) [inline packages/pages/src/components/image/image.tsx:14]: Did we consider making the prop just `document.logo` to simplify things a bit?

Comment (EmilyZhang777) [reply packages/pages/src/components/image/image.tsx:14]: If we want to make this change, the `ImageType` will need to be `type ImageType = {image: {height: number...};};` with the extra `image` layer.

File: packages/pages/src/components/image/image.tsx [hunk 7]
```diff
@@ -1,14 +1,34 @@
 import * as React from "react";
-import { useRef, useState, useEffect } from "react";
-import { ImageProps } from "./types";
-
+import { useEffect, useRef, useState } from "react";
+import { ImageProps, ImageLayout } from "./types";
+
+/**
+ * A component that renders an image. Here is an example of using the component with simple and
```
Comment (mkilpatrick) [inline packages/pages/src/components/image/image.tsx:6]: I'd rephrase this intro to:

Renders an image based from the Yext Knowledge Graph

File: packages/pages/src/components/image/image.tsx (hunk 6 above)
Comment (mkilpatrick) [reply packages/pages/src/components/image/image.tsx:14]: I agree with Max that we should make it simpler to use.

File: packages/pages/src/components/image/image.tsx [hunk 8]
```diff
@@ -1,14 +1,34 @@
...
+ * complex image fields from Knowledge Graph:
+ * ```
+ * import { Image } from "@yext/pages/components";
+ *
+ * const template = ({ document }) => {
```
Comment (mkilpatrick) [inline packages/pages/src/components/image/image.tsx:11]: I wouldn't give an example of the whole template, just using the component itself.

File: packages/pages/src/components/image/image.tsx [hunk 9]
```diff
@@ -1,14 +1,34 @@
...
+ *   return (
+ *     <div>
+ *       <Image image={document.logo.image} />
+ *       <Image image={document.photoGallery[0].image} />
+ *     </div>
+ *   );
+ * }
+ * ```
+ *
+ * `layout`, `imgOverrides` and `style` can be used to specify the style and behavior of the image.
```
Comment (mkilpatrick) [inline packages/pages/src/components/image/image.tsx:21]: I think this is better left to documenting the prop type itself, rather than the component.

File: packages/pages/src/components/image/image.tsx [hunk 10]
```diff
@@ -22,24 +42,78 @@ export const Image = ({
...
       "Width or height is passed in but layout is not fixed. These will have no impact. If you want to have a fixed height or width then set layout to fixed."
     );
   }
 
-  const imgWidth: number = image.image.width;
-  const imgHeight: number = image.image.height;
-  const imgUUID: string = image.image.url.split("/")[4];
+  const imgWidth: number = image.width;
+  const imgHeight: number = image.height;
+  const imgUrlValid = !!image.url && image.url.split("/").length > 4;
```
Comment (mkilpatrick) [inline packages/pages/src/components/image/image.tsx:53]: Turn this validation into its own function. I also think we should validate it more that just checking there are 4 slashes. We should also check our mktgcdn domain.

File: packages/pages/src/components/image/image.tsx [hunk 11]
```diff
@@ -22,24 +42,78 @@ export const Image = ({
...
+        srcSet={srcSet}
+        loading={"lazy"}
+        onLoad={() => setImgLoaded(true)}
+        {...imgOverrides}
+      />
+    </>
+  );
+};
 
+const getImageUrl = (uuid: string, width: number, height: number) => {
```
Comment (mkilpatrick) [inline packages/pages/src/components/image/image.tsx:94]: Add jsdoc to these, even though they're internal

File: packages/pages/src/components/image/types.ts [hunk 12]
```diff
@@ -7,22 +7,52 @@ export type ThumbnailType = {
 };
 
 export type ImageType = {
```
Comment (mkilpatrick) [inline packages/pages/src/components/image/types.ts:9]: Sounds like we want the double image here

File: packages/pages/package.json [hunk 13]
```diff
@@ -115,43 +113,32 @@
...
+      "<rootDir>/pages/src/**/*.test.{ts,tsx,js}"
     ],
     "transform": {
       "\\.[jt]sx?$": "ts-jest"
     },
     "moduleNameMapper": {
       "^(.+)\\.js$": "$1"
-    }
+    },
+    "testEnvironment": "jsdom"
```
Comment (mkilpatrick) [inline packages/pages/package.json:142]: I don't think we want to globally change this. The default test env is node. If you need jsdom for your specific test then you can override it for your particular test suite. See `packages/pages/src/util/runtime.test.ts`.

File: packages/pages/src/components/image/image.test.tsx [hunk 14]
```diff
@@ -0,0 +1,219 @@
...
+const image = {
+  image: {
+    width: imgWidth,
+    height: imgHeight,
+    url: `https://a.mktgcdn.com/p/${imgUUID}/2x1.jpg`,
+  },
+};
+
+/**
+ * @jest-environment node
```
Comment (mkilpatrick) [inline packages/pages/src/components/image/image.test.tsx:21]: Same comment. Why did you set the env to jsdom globally and then just swap it back here? Also, I'm pretty sure this needs to be at the top of the file to work.

File: packages/pages/src/components/image/image.test.tsx [hunk 15]
```diff
@@ -0,0 +1,219 @@
...
+ */
+
+describe("Image", () => {
+  it(`properly logs warning when layout is not ${ImageLayout.FIXED} and width or height is provided`, async () => {
+    jest.clearAllMocks();
```
Comment (mkilpatrick) [inline packages/pages/src/components/image/image.test.tsx:26]: Shouldn't be needed at the beginning if your other tests properly clear at the end

File: packages/pages/src/components/image/image.test.tsx [hunk 16]
```diff
@@ -0,0 +1,219 @@
...
+      image={image}
+      layout={ImageLayout.INTRINSIC}
+      width={100}
+    />);
+    expect(logMock.mock.calls.length).toBe(1);
+
+    jest.clearAllMocks();
+  });
+
+  it("properly renders the image with the pass through style and imgOverrides", () => {
```
Comment (mkilpatrick) [inline packages/pages/src/components/image/image.test.tsx:40]: Is there a way to test the lazy loading-ness of the image?

File: packages/pages/src/components/image/image.tsx [hunk 17]
```diff
@@ -91,21 +79,43 @@ export const Image = ({
   );
 };
 
-const getImageUrl = (uuid: string, width: number, height: number) => {
+/**
+ * Returns the UUID of an image given its url. Logs a warning message if the image url is invalid.
+ */
+export const getImageUUID = (url: string) => {
```
Comment (mkilpatrick) [inline packages/pages/src/components/image/image.tsx:85]: Does this need to be exported?

File: packages/pages/src/components/image/image.tsx [hunk 18]
```diff
@@ -91,21 +79,43 @@ export const Image = ({
...
+  const uuidRegex = /(?<=^https:\/\/a\.mktgcdn\.com\/p\/)[a-zA-Z0-9]+(?=\/(.*)$)/g;
```
Comment (mkilpatrick) [inline packages/pages/src/components/image/image.tsx:86]: Move this outside of the func and call it MKTGCDN_URL_REGEX

File: packages/pages/src/vite-plugin/build/buildStart/rendering/templateUtils.test.tsx [hunk 19]
```diff
@@ -16,12 +16,24 @@ const baseTemplateModule: TemplateModuleInternal<any, any> = {
 
 const baseProps: TemplateProps = {
   document: {},
-  __meta: { mode: "development" },
+  __meta: {
+    mode: "development",
+    manifest: {
```
Comment (mkilpatrick) [inline packages/pages/src/vite-plugin/build/buildStart/rendering/templateUtils.test.tsx:21]: Did this need to change?

File: packages/pages/src/components/image/image.test.tsx (hunk 14 above)
Comment (EmilyZhang777) [reply packages/pages/src/components/image/image.test.tsx]: Didn't get this to work and forgot to remove it lol

File: packages/pages/src/components/image/image.tsx (hunk 17 above)
Comment (EmilyZhang777) [reply packages/pages/src/components/image/image.tsx:87]: I exported them for testing but they are not exported in index.ts

File: packages/pages/src/vite-plugin/build/buildStart/rendering/templateUtils.test.tsx (hunk 19 above)
Comment (EmilyZhang777) [reply packages/pages/src/vite-plugin/build/buildStart/rendering/templateUtils.test.tsx:21]: yeah, need to change this to get the test ran properly

File: packages/pages/src/components/image/image.test.tsx (hunk 16 above)
Comment (EmilyZhang777) [reply packages/pages/src/components/image/image.test.tsx:38]: I don't think so. The onLoad on the img tag is not invoked in test so there's no good way to test the loading option

File: packages/pages/package.json [hunk 20]
```diff
@@ -90,7 +90,7 @@
     "@types/express-serve-static-core": "^4.17.29",
     "@types/fs-extra": "^9.0.13",
     "@types/glob": "^7.2.0",
-    "@types/jest": "^27.5.2",
+    "@types/jest": "^27.5.1",
```
Comment (mkilpatrick) [inline packages/pages/package.json:93]: You actually moved the version down which seems weird, though it doesn't actually matter.

Comment (EmilyZhang777) [reply packages/pages/package.json:93]: It was at 27.5.1 before. I thought. the check failure was caused by this so gave it a try. 27.5.1 should be fine to use

File: packages/pages/src/components/image/image.tsx [hunk 21]
```diff
@@ -0,0 +1,210 @@
...
+};
+
+/**
+ * Returns the UUID of an image given its url. Logs a warning message if the image url is invalid.
+ */
+export const getImageUUID = (url: string) => {
+  const matches = url.match(MKTGCDN_URL_REGEX);
+
+  if (matches == null || matches.length == 0) {
+    console.warn(`Invalid image url: ${url}.`);
```
Comment (jimboHenris) [inline packages/pages/src/components/image/image.tsx:109]: is this a recoverable state? If not, we should log a full error

File: packages/pages/src/components/image/image.tsx [hunk 22]
```diff
@@ -0,0 +1,210 @@
...
+
+      break;
+    case ImageLayout.FIXED:
+      if (!width && !height) {
+        console.warn(
+          "Using fixed layout but width and height are not passed as props."
+        );
+      }
+
+      const fixedWidth = width
```
Comment (jimboHenris) [inline packages/pages/src/components/image/image.tsx:161]: as a general rule, I try to avoid nested ternary operators, I use tao of react to guide most of my style / design decisions and this is something that's mentioned: https://alexkondov.com/tao-of-react/#avoid-nested-ternary-operators

File: packages/pages/src/components/image/image.tsx [hunk 23]
```diff
@@ -0,0 +1,210 @@
...
+    .join(", ");
+
+  const onError = () => {
+    setImgStatus(imgLoadingStatus.ERROR);
+    console.warn(`Invalid image src: ${src}.`);
+  };
+
+  return (
+    <>
+      {placeholder != null &&
```
Comment (jimboHenris) [inline packages/pages/src/components/image/image.tsx:80]: Just to be certain there's never an instance in which both placeholder and  could render I'd recommend just switching on imgLoadingStatus.LOADED. If it's loaded, , otherwise return placeholder. I don't think it'll be an issue if it's null, right? That's equivalent to what'd happen if we were in an error state w/o a placeholder anyway I think

File: packages/pages/src/components/image/image.tsx [hunk 24]
```diff
@@ -0,0 +1,210 @@
...
+      // Don't let image be wider than its intrinsic width
+      style.maxWidth = imgWidth;
+      style.width = "100%";
+      style.aspectRatio = aspectRatio
+        ? `${aspectRatio}`
+        : `${imgWidth} / ${imgHeight}`;
+
+      break;
+    case ImageLayout.FIXED:
+      if (!width && !height) {
```
Comment (jimboHenris) [inline packages/pages/src/components/image/image.tsx:155]: I'd move all this validation into a helper function, combine with the check at line 47 and call it there

File: packages/pages/src/components/image/types.ts [hunk 25]
```diff
@@ -0,0 +1,60 @@
...
+  width?: number;
+  /** The absolute height of the image. Only impacts if layout is set to "fixed". */
+  height?: number;
+  /** The aspect ratio of the image. Only impacts if layout is set to "aspect". */
+  aspectRatio?: number;
+  /** Specifies how the image is rendered. */
+  layout?: ImageLayout;
+  /** A pass through react component that is displayed when the image is loading. */
+  placeholder?: React.ReactNode;
+  /** Pass through props that are on the native HTML img tag. The Image component may not work if src and/or srcsets are included. */
```
Comment (jimboHenris) [inline packages/pages/src/components/image/types.ts:56]: can we just add a check and if there's both ignore src and use srcset?

File: packages/pages/src/components/image/image.tsx [hunk 26]
```diff
@@ -0,0 +1,210 @@
...
+      const fixedHeight = height
+        ? height
+        : width
+        ? (width * imgHeight) / imgWidth
+        : imgHeight;
+
+      style.width = fixedWidth;
+      style.height = fixedHeight;
+
+      if (fixedWidth && fixedHeight) {
```
Comment (jimboHenris) [inline packages/pages/src/components/image/image.tsx:175]: what's the case where these both wouldn't be set here?

File: packages/pages/src/components/image/image.tsx (hunk 23 above)
Comment (EmilyZhang777) [reply packages/pages/src/components/image/image.tsx:80]: Not sure if I get your point. But the main issue with this is, we want to make sure  is rendered in order to trigger onLoad, which sets the imgStatus. I verified that the  and placeholder never show up together

File: packages/pages/src/components/image/image.tsx [hunk 27]
```diff
@@ -3,7 +3,12 @@ import { useEffect, useRef, useState } from "react";
 import { ImageProps, ImageLayout } from "./types";
 
 const MKTGCDN_URL_REGEX =
-  /(?<=^https:\/\/a\.mktgcdn\.com\/p\/)[a-zA-Z0-9]+(?=\/(.*)$)/g;
+  /((?<=^http:\/\/a\.mktgcdn\.com\/p\/)|(?<=^https:\/\/a\.mktgcdn\.com\/p\/)).+(?=\/(.*)$)/g;
+enum imgLoadingStatus {
```
[the remark this answers is not in the document]
Comment (mkilpatrick) [reply packages/pages/src/components/image/image.tsx:7]: Actually I don't think we need this at all. I'm going to put up a commit with some changes. TAL and lmk what you think.

File: packages/pages/src/components/image/image.tsx (hunk 23 above)
Comment (mkilpatrick) [reply packages/pages/src/components/image/image.tsx]: Jim, I think the issue here is that the placeholder might not exist. We only display it if there is one.

File: packages/pages/src/components/image/types.ts (hunk 25 above)
Comment (EmilyZhang777) [reply packages/pages/src/components/image/types.ts:58]: The reason that this might not work is because we set the src and srcset on the ``. I don't think we reached a conclusion on which src has the higher priority. I'll leave it as is for now. If we want to add support for it, it can be a future item 👍

File: packages/pages/src/components/image/image.tsx (hunk 26 above)
Comment (EmilyZhang777) [reply packages/pages/src/components/image/image.tsx]: I changed the behavior and forgot to remove this check

File: packages/pages/src/components/image/image.tsx [hunk 28]
```diff
@@ -70,47 +73,89 @@ export const Image = ({
...
+  if (imgWidth < 0) {
+    console.warn(`Invalid image width: ${imgWidth}.`);
+  }
+
+  if (imgHeight < 0) {
+    console.warn(`Invalid image height: ${imgHeight}.`);
+  }
+
+  if (layout == ImageLayout.FIXED) {
+    if (!width && !height) {
```
Comment (mkilpatrick) [inline packages/pages/src/components/image/image.tsx:115]: Is height actually required? When I was testing earlier a fixed image with just a width seemed fine.

File: packages/pages/src/components/image/image.tsx [hunk 29]
```diff
@@ -34,34 +28,43 @@ export const Image = ({
...
+  const imgWidth: number = Math.abs(image.image.width);
+  const imgHeight: number = Math.abs(image.image.height);
   const imgUUID = getImageUUID(image.image.url);
 
+  // The image is invalid, only try to load the placeholder
+  if (!imgUUID) {
+    return <>{placeholder != null && placeholder}</>;
+  }
+
+  const absWidth = width ? Math.abs(width) : undefined;
```
Comment (mkilpatrick) [inline packages/pages/src/components/image/image.tsx:57]: Why is this needed? Validation should handle it. It's confusing to validate something, then also fix it and use some other variable here.

File: packages/pages/src/components/image/image.tsx [hunk 30]
```diff
@@ -70,47 +73,89 @@ export const Image = ({
...
-  };
-
   return (
     <>
-      {placeholder != null &&
-        imgStatus != imgLoadingStatus.LOADED &&
-        placeholder}
-      {imgStatus != imgLoadingStatus.ERROR && (
+      {!imgLoaded && placeholder != null && placeholder}
+      {src && (
```
Comment (mkilpatrick) [inline packages/pages/src/components/image/image.tsx:79]: I don't think the src check is necessary. It's always set by the time we get here.

File: packages/pages/src/components/image/image.test.tsx [hunk 31]
```diff
@@ -125,101 +148,172 @@ describe("handleLayout", () => {
...
     );
 
-    expect(imgStyle.aspectRatio).toEqual(aspectRatio.toString());
+    expect(logMock.mock.calls.length).toBe(1);
+    jest.clearAllMocks();
   });
 
-  it(`properly sets width when layout is ${ImageLayout.FILL}`, () => {
-    const { imgStyle } = handleLayout(
+  it(`properly logs warning when image.width is a negative value`, () => {
```
[the remark this answers is not in the document]
Comment (EmilyZhang777) [reply packages/pages/src/components/image/image.test.tsx:275]: nah, this is the ImageProps.image.width, the other one is for ImageProps.width

File: packages/pages/src/components/image/image.tsx [hunk 32]
```diff
@@ -132,8 +177,8 @@ export const handleLayout = (
   imgHeight: number,
   imgUUID: string,
   style: React.CSSProperties,
-  width?: number,
-  height?: number,
+  absWidth?: number,
+  absHeight?: number,
   aspectRatio?: number
 ): { src: string; imgStyle: React.CSSProperties; widths: number[] } => {
   let widths: number[] = [100, 320, 640, 960, 1280, 1920];
```
[the remark this answers is not in the document]
Comment (EmilyZhang777) [reply packages/pages/src/components/image/image.tsx:184]: It's not really mutating, this function takes in the style prop the user passed in and returns a new imgStyle that will be used on the ``

File: packages/pages/src/components/image/image.tsx [hunk 33]
```diff
@@ -70,47 +73,89 @@ export const Image = ({
...
+export const validateRequiredProps = (
+  layout: ImageLayout,
+  imgWidth: number,
+  imgHeight: number,
+  width?: number,
+  height?: number,
+  aspectRatio?: number
+) => {
+  if (imgWidth < 0) {
+    console.warn(`Invalid image width: ${imgWidth}.`);
```
[the remark this answers is not in the document]
Comment (EmilyZhang777) [reply packages/pages/src/components/image/image.tsx:107]: This is not a deadly error, we can still render the image. All cases that log a warning now can still render an image. The only cases that we can't render an image are invalid url and invalid uuid, which are logged as errors.

File: packages/pages/src/components/image/image.tsx (hunk 28 above)
Comment (EmilyZhang777) [reply packages/pages/src/components/image/image.tsx:115]: we need at least a width or a height to render

File: packages/pages/src/components/image/image.tsx (hunk 29 above)
Comment (EmilyZhang777) [reply packages/pages/src/components/image/image.tsx:57]: hmm, the same logic is used throughout the file (where even if you have an invalid config, we still try to render an image to our best effort. For example, if you have layout = "aspect" but do not provide a ImageProps.aspectRatio, we still try to render an image with the `aspectRatio = ImageProps.image.width/ImageProps.image.height`.  Do we want to change that?

Comment (mkilpatrick) [reply packages/pages/src/components/image/image.tsx]: This is probably a call for Max. Idk if "try to render it as best we can" is good or if it leads to unknown outcomes. We can go with "try our best" for now, but can you start a thread about it (tomorrow is fine) and cc me?

File: packages/pages/src/components/image/image.test.tsx [hunk 34]
```diff
@@ -66,16 +66,46 @@ describe("Image", () => {
...
         image={{
-          image: { ...image.image, url: "https://a.mktgcdn.com/p/2x1.jpg" },
+          image: { ...image.image, url: invalidUrl },
         }}
         placeholder={placeholder}
       />
     );
 
     expect(screen.getByText(placeholderText)).toBeTruthy();
+
```
Comment (mkilpatrick) [inline packages/pages/src/components/image/image.test.tsx:84]: remove empty line

File: packages/pages/src/components/image/image.tsx (hunk 32 above)
[the remark this answers is not in the document]
Comment (mkilpatrick) [reply packages/pages/src/components/image/image.tsx:182]: Right, you return a new imgStyle, but you mutate style throughout first

File: packages/pages/src/components/image/image.tsx (hunk 33 above)
[the remark this answers is not in the document]
Comment (mkilpatrick) [reply packages/pages/src/components/image/image.tsx:105]: Sounds good assuming we're trying to render something as best we can. Thanks for the explanation.

File: packages/pages/src/components/image/image.tsx (hunk 28 above)
Comment (mkilpatrick) [reply packages/pages/src/components/image/image.tsx:113]: Oh yea, ha, misread. 😓

**Review by mkilpatrick — APPROVED** [review]

File: packages/pages/src/components/image/image.tsx [hunk 35]
```diff
@@ -0,0 +1,271 @@
...
+  width,
+  height,
+  aspectRatio,
+  layout = ImageLayout.INTRINSIC,
+  placeholder,
+  imgOverrides,
+  style = {},
+}: ImageProps) => {
+  const imgRef = useRef<HTMLImageElement>(null);
+  const [imgLoaded, setImgLoaded] = useState(false);
```
Comment (mancusi) [inline packages/pages/src/components/image/image.tsx:31]: isImageLoaded, setIsImageLoaded

File: packages/pages/src/components/image/image.tsx [hunk 36]
```diff
@@ -0,0 +1,271 @@
...
+
+  useEffect(() => {
```
Comment (mancusi) [inline packages/pages/src/components/image/image.tsx:33]:
```js
const useIsImageLoaded = (): [React.Ref<HTMLImageElement>, boolean] => {
    const imgRef = useRef<HTMLImageElement>(null);
    const [imgLoaded, setImgLoaded] = useState(false);
    
    useEffect(() => {
       if (imgRef.current?.complete) {
      setImgLoaded(true);
    }
  }, []);

  return imgRef, isImageLoaded;
}
```

Comment (mancusi) [reply packages/pages/src/components/image/image.tsx:33]: then once this is a custom hook move it outside the view;

File: packages/pages/src/components/image/image.tsx [hunk 37]
```diff
@@ -0,0 +1,271 @@
...
+      <img
+        ref={imgRef}
+        style={imgStyle}
+        src={src}
+        className={className}
+        width={absWidth}
+        height={absHeight}
+        srcSet={srcSet}
+        loading={"lazy"}
+        onLoad={() => setImgLoaded(true)}
```
Comment (mancusi) [inline packages/pages/src/components/image/image.tsx:88]: apparently load doesn't trigger when the image is loaded in chrome so I think just checking complete is fine.

**Review by mancusi — CHANGES_REQUESTED** [review]

File: packages/pages/src/components/image/types.ts [hunk 38]
```diff
@@ -0,0 +1,62 @@
...
+   * `objectPosition` props of the `style` rpop.
+   */
+  FIXED: "fixed",
+  /** Shows the image in a fixed aspect ratio. The `aspectRatio` prop must be provided. */
+  ASPECT: "aspect",
+  /** Always fills the image to 100% of the container's width. */
+  FILL: "fill",
+} as const;
+
+export type ImageLayout = typeof ImageLayout[keyof typeof ImageLayout];
```
Comment (mancusi) [inline packages/pages/src/components/image/types.ts:37]: any exported value should have some sort of documentation on it. same as above

File: packages/pages/src/components/image/types.ts [hunk 39]
```diff
@@ -0,0 +1,62 @@
...
+export type ImageType = {
+  image: {
+    height: number;
+    width: number;
+    url: string;
+    thumbnails?: ThumbnailType[];
+  };
+};
+
+export const ImageLayout = {
```
Comment (mancusi) [inline packages/pages/src/components/image/types.ts:18]: this and the type have the same name which isn't ideal -- maybe this can be called ImageLayoutOption?

File: packages/pages/src/components/image/image.tsx (hunk 37 above)
Comment (mancusi) [reply packages/pages/src/components/image/image.tsx:88]: nvm this isn't true (just tested)

File: packages/pages/src/components/image/image.tsx (hunk 36 above)
Comment (mancusi) [reply packages/pages/src/components/image/image.tsx:33]: nvm we can't do this as nicely since we have to call the setter in the img component itself

File: packages/pages/src/components/image/image.tsx (hunk 29 above)
Comment (EmilyZhang777) [reply packages/pages/src/components/image/image.tsx]: looks like Max wants to go with the current way

File: packages/pages/src/components/image/types.ts [hunk 40]
```diff
@@ -34,29 +49,58 @@ export const ImageLayout = {
...
   /** A pass through react component that is displayed when the image is loading. */
   placeholder?: React.ReactNode;
   /** Pass through props that are on the native HTML img tag. The Image component may not work if src and/or srcsets are included. */
   imgOverrides?: Object;
   /** The pass through style of the underlying img tag. */
   style?: React.CSSProperties;
-};
+}
+
+interface OtherImageProps extends BaseImageProps {
```
Comment (mkilpatrick) [inline packages/pages/src/components/image/types.ts:81]: jsdoc for all of these with @public

File: packages/pages/src/components/image/types.ts [hunk 41]
```diff
@@ -34,29 +49,58 @@ export const ImageLayout = {
...
+interface AspectImageProps extends BaseImageProps {
+  /** Specifies how the image is rendered. */
+  layout: "aspect";
+  /** The aspect ratio of the image. Only impacts if layout is set to "aspect". */
+  aspectRatio: number;
+}
+
+/**
+ * The shape of the data passed directly to the different template functions with the
+ * exception of the render function (getPath, getHeadConfig, etc).
```
Comment (mkilpatrick) [inline packages/pages/src/components/image/types.ts:104]: This doc talks about potential usage rather than what it _is_. These are the props required by the Image componet.

File: packages/pages/src/components/image/types.ts (hunk 40 above)
Comment (EmilyZhang777) [reply packages/pages/src/components/image/types.ts:81]: I don't think we want to export these?

Comment (EmilyZhang777) [reply packages/pages/src/components/image/types.ts:88]: Added docs but didn't export the layout specific interfaces. Let me know if you want to export them.

**Comment by jimboHenris** [comment]
Aside from the failing GH actions, LGTM

**Review by mancusi — APPROVED** [review]
forgot to approve this earlier 👍

**Review by mkilpatrick — APPROVED** [review]
